### PR TITLE
[SPARK-7330] [SQL] avoid NPE at jdbc rdd

### PR DESCRIPTION
**Pull request by adrian-wang**
Thank @nadavoosh point this out in #5590

### CODE REVIEW

**Comment by adrian-wang** [comment]
retest this please.

**Comment by adrian-wang** [comment]
retest this please.

**Comment by adrian-wang** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala [hunk 1]
```diff
@@ -361,8 +361,14 @@ private[sql] class JDBCRDD(
           val pos = i + 1
           conversions(i) match {
             case BooleanConversion    => mutableRow.setBoolean(i, rs.getBoolean(pos))
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala]: will other data types also have the NPE problem?

Comment (adrian-wang) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala]: For now only Date will have the NPE problem. in #5456 when I convert the decimal type there I also apply similar approach when I call `def Decimal(x: BigDecimal)`.
I have add another test case to make sure it works for all types.

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala]: Do we need to handle null value for StringType and TimestampType?

Comment (adrian-wang) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala]: We don't need to. After this part we will call setNullAt for those null values.

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala]: OK i see. Can we add a comment at here to explain that `DateUtils.fromJavaDate` does not handle `null` input?

Comment (adrian-wang) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala]: @yhuai done.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 2]
```diff
@@ -266,6 +284,18 @@ class JDBCSuite extends FunSuite with BeforeAndAfter {
...
+    val rows = TestSQLContext.jdbc(urlWithUserAndPass, "TEST.NULLTYPES").collect()
+    assert(rows(0).isNullAt(0))
+    assert(rows(0).isNullAt(1))
+    assert(rows(0).isNullAt(2))
+    assert(rows(0).isNullAt(3))
+    assert(rows(0).isNullAt(4))
+    assert(rows(0).isNullAt(5))
+    assert(rows(0).isNullAt(6))
+    assert(rows(0).isNullAt(7))
+  }
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: For this test, can we test all data types?

**Comment by yhuai** [comment]
Can you resolve the conflict?

**Comment by adrian-wang** [comment]
@yhuai sure, I will rebase that

**Comment by adrian-wang** [comment]
retest this please.

**Comment by shaneknapp** [comment]
i'll retrigger this build once jenkins is back up...

**Comment by shaneknapp** [comment]
jenkins, test this please

**Comment by yhuai** [comment]
LGTM. I am merging it to master and branch-1.4.

**Comment by yhuai** [comment]
Also branch-1.3.

**Pull request closed without merging**